File: packages/adapter-erb/lib/barefoot_js/search_params.rb
# frozen_string_literal: true

module BarefootJS
  # Request-scoped SSR view of the query string behind the reactive
  # `searchParams()` environment signal (router v0.5, #1922). The framework
  # integration builds one per request from the request URL and threads it
  # into the template scope as `v[:searchParams]` (the camelCase JS name the
  # adapters keep, like every other signal/prop var); the compiled ERB
  # template reads it via `v[:searchParams].get('key')`.
  #
  # This runtime is template-engine- and framework-agnostic (Ruby stdlib
  # only), matching the rest of BarefootJS, so it can ship standalone.
  #
  # Semantics mirror the browser's URLSearchParams#get exactly under the
  # adapters' `?? -> ||=`-style lowering: get() returns the first value for a
  # key, or `nil` when the key is absent. A present-but-empty value
  # (`?sort=`) keeps the empty string -- the same distinction JS `??` draws
  # between `null` and `''`.
  class SearchParams
    # Parse a raw query string into the reader. A leading '?' is tolerated,
    # '+' decodes to a space, and %XX escapes are decoded -- mirroring
    # URLSearchParams's application/x-www-form-urlencoded parsing. A
    # malformed pair never raises; it simply contributes nothing, matching
    # the browser's lenient parsing.
    def initialize(query = '')
      query ||= ''
      query = query.sub(/\A\?/, '')
      @values = Hash.new { |h, k| h[k] = [] }
      query.split(/[&;]/).each do |pair|
        next if pair.empty?

        key, val = pair.split('=', 2)
        key = decode(key)
        val = val.nil? ? '' : decode(val)
        @values[key] << val
      end
    end

    # First value for `key`, or `nil` when the key is absent (see the class
    # docstring for why `nil` -- not '' -- is the right "missing" sentinel).
    # A present-but-empty value returns ''.
    def get(key)
      vals = @values[key]
      return nil if vals.nil? || vals.empty?

      vals.first
    end

    private

    def decode(s)
      s = (s || '').dup
      s.tr!('+', ' ')
      # %XX -> raw octet, then interpret the octet stream as UTF-8 (what
      # URLSearchParams does). A byte run that isn't valid UTF-8 is
      # replaced rather than raising (lenient parsing, mirrors the Perl
      # port's leave-as-is policy as closely as Ruby's String#encode allows).
      bytes = s.gsub(/%([0-9A-Fa-f]{2})/) { [Regexp.last_match(1)].pack('H2') }
      bytes.force_encoding('UTF-8')
      bytes.valid_encoding? ? bytes : bytes.force_encoding('ASCII-8BIT').encode('UTF-8', invalid: :replace, undef: :replace)
    end
  end
end

File: packages/adapter-pebble/java/src/main/java/dev/barefootjs/pebble/SearchParams.java
package dev.barefootjs.pebble;

import java.nio.charset.StandardCharsets;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

/**
 * Java port of `packages/adapter-jinja/python/barefootjs/search_params.py`
 * (itself ported from `packages/adapter-perl/lib/BarefootJS/SearchParams.pm`)
 * — the request-scoped SSR view of the query string behind the reactive
 * `searchParams()` environment signal (#1922). The compiled `.peb` template
 * reads it as `{{ searchParams.get('key') }}` — Pebble resolves a
 * dot-method call against any public method on the context value, so this
 * needs no adapter-side special casing beyond seeding a `searchParams`
 * key of this type in the render context (see {@link Main}).
 *
 * <p>Semantics mirror the browser's `URLSearchParams.get` exactly: the
 * first value for a key, or `null` when the key is absent (a
 * present-but-empty value returns `""`, distinct from absent — the `??`
 * lowering's null/undefined-only coalescing relies on this distinction).
 */
final class SearchParams {

  private final Map<String, List<String>> values = new LinkedHashMap<>();

  SearchParams(String query) {
    String q = query == null ? "" : query;
    if (q.startsWith("?")) {
      q = q.substring(1);
    }
    if (q.isEmpty()) {
      return;
    }
    for (String pair : q.split("[&;]", -1)) {
      if (pair.isEmpty()) {
        continue;
      }
      int eq = pair.indexOf('=');
      String key = eq < 0 ? pair : pair.substring(0, eq);
      String val = eq < 0 ? null : pair.substring(eq + 1);
      key = decode(key);
      String decodedVal = val == null ? "" : decode(val);
      values.computeIfAbsent(key, k -> new java.util.ArrayList<>()).add(decodedVal);
    }
  }

  /** First value for `key`, or `null` when the key is absent. */
  public String get(String key) {
    List<String> vals = values.get(key);
    return (vals == null || vals.isEmpty()) ? null : vals.get(0);
  }

  /** `application/x-www-form-urlencoded` decode: `+` -> space, then percent-decode. Never throws on malformed input (lenient, matching the browser). */
  private static String decode(String s) {
    String withSpaces = s.replace('+', ' ');
    java.io.ByteArrayOutputStream raw = new java.io.ByteArrayOutputStream();
    int n = withSpaces.length();
    int i = 0;
    while (i < n) {
      char c = withSpaces.charAt(i);
      if (c == '%' && i + 2 < n && isHex(withSpaces.charAt(i + 1)) && isHex(withSpaces.charAt(i + 2))) {
        raw.write(Integer.parseInt(withSpaces.substring(i + 1, i + 3), 16));
        i += 3;
      } else {
        byte[] bytes = String.valueOf(c).getBytes(StandardCharsets.UTF_8);
        raw.write(bytes, 0, bytes.length);
        i += 1;
      }
    }
    return new String(raw.toByteArray(), StandardCharsets.UTF_8);
  }

  private static boolean isHex(char c) {
    return (c >= '0' && c <= '9') || (c >= 'a' && c <= 'f') || (c >= 'A' && c <= 'F');
  }
}
